### PR TITLE
Difficult track iteration rewrites

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #4957

Depends on
- #4908

Some iterations with Any and Selected are changed to Leaders and SelectedLeaders,
by lowering looping over channels into member functions of WaveTrack.

Some member functions of WaveTrack and virtual member functions of Track that
produced other track objects are changed to return lists of tracks, and Paste is changed to
accept a list of tracks as an argument.  This change of functions signatures may be undone
in the future when this great transformation is completed, and wide wave tracks really exist
as single tracks.

See commits 56175dd3270c30ded5a69e48d8d94e9e642dcde6 and
ceb695e8576a1abfc9aae7b06729e205a2697ea1 which already made similar changes for
two member functions of WaveTrack.

QA: 
Known BUG introduced by this, to be repaired when we really have "wide" Wave clips:
- #4954

Test effects, specifically selecting only a part of a mono or stereo track, including multiple clips, with and without gaps between:
- [x] Change Speed
- [x] Noise or tone or chirp generator
- [x] Noise Reduction
- [x] Change Pitch (high and low quality)
- [x] Change Tempo (high and low quality)
- [x] Sliding stretch
- [x] Any Nyquist effect
- [x] Equalization (test either of the two UI)
- [x] EQ and Compressor applied to tracks that are not in 32 bit float format
- [x] PaulStretch
- [x] Stereo To Mono
- [x] Truncate Silence
- [x] Repeat

Also these editing operations, on mono and stereo tracks:
- [x] Cut (all kinds of tracks, and with cutlines enabled in Preferences)
- [x] Copy (all kinds)
- [x] Split Cut
- [x] Split New
- [x] Split Delete
- [x] Paste (of all kinds of tracks, and of multiple copied clips)
- [x] Cross-project paste (all kinds of tracks, and including hidden smart clip data, and into an empty project)
- [x] Trim (Ctrl + T)
- [x] Duplicate (Ctrl + D) (all kinds)
- [x] Delete (Ctrl + K)
- [x] All the edit-by-labels commands (see #3679; only preserving behavior, not fixing the bugs; you may need sync lock on)
- [x] Detach at Silences
- [x] Silence (Ctrl + L) (not the generator)
- [x] Split
- [x] Join
- [x] Effect preview
- [x] Sync-lock adjustment of tracks of all kinds, to shorter or longer length
- [x] Select in a wave track, then generate Silence of 0 duration
- [x] Punch and Roll recording
- [x] Recording into existing or same track; with view changes and Ctrl + B to drop labels during recording; then undo and redo
- [x] Undo and redo of changes to all kinds of tracks
- [x] Make a stereo track; pick in it and run a Nyquist generator
- [x] Recording to new track that would be off-screen correctly scrolls the view vertically
- [x] File > Export as ... commands correctly make mono or stereo files depending on the tracks in the project
- [x] Persistency of mono and stereo tracks
- [x] Find Zero Crossings in mono and stereo
- [x] Plot Spectrum for mono and stereo (just use default settings)
- [x] Changing WaveTrack color (in the context menu); and mix and render preserves color; mono and stereo
- [x] Set Track Visuals macro command to change color
- [x] Conversion of track sample format (mono and stereo)

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by Paul-Licameli** [comment]
Rebased onto updated Rewrite-some-effects branch with no changes

**Comment by Paul-Licameli** [comment]
Rebased onto recent master to resolve the conflict in NoiseReduction.cpp

**Comment by Paul-Licameli** [comment]
One new bug fix commit for review

File: src/menus/SelectMenus.cpp [hunk 1]
```diff
@@ -38,65 +38,72 @@ double NearestZeroCrossing
    Floats dist{ windowSize, true };
 
    int nTracks = 0;
-   for (auto one : tracks.Selected< const WaveTrack >()) {
+   for (auto one : tracks.Leaders<const WaveTrack>()) {
```
Comment (crsib) [inline src/menus/SelectMenus.cpp:41]: Should this be SelectedLeaders?

File: libraries/lib-effects/MixAndRender.cpp [hunk 2]
```diff
@@ -19,17 +19,15 @@ Paul Licameli split from Mix.cpp
 using WaveTrackConstArray = std::vector < std::shared_ptr < const WaveTrack > >;
 
 //TODO-MB: wouldn't it make more sense to DELETE the time track after 'mix and render'?
-void MixAndRender(const TrackIterRange<const WaveTrack> &trackRange,
+TrackListHolder MixAndRender(const TrackIterRange<const WaveTrack> &trackRange,
```
Comment (crsib) [inline libraries/lib-effects/MixAndRender.cpp:22]: I vote this for the best commit of this PR :-)

File: src/effects/Equalization.cpp [hunk 3]
```diff
@@ -430,24 +440,20 @@ bool EffectEqualization::Process(EffectInstance &, EffectSettings &)
...
-            if (ProcessOne(task, count, *pChannel, start, len)) {
-               temp->Add(pNewChannel);
-               assert(pNewChannel->IsLeader() == pChannel->IsLeader());
-            }
-            else
-               bGoodResult = false;
+
+            bGoodResult = ProcessOne(task, count, *pChannel, start, len);
+            if (!bGoodResult)
+               goto done;
```
Comment (crsib) [inline src/effects/Equalization.cpp:448]: I don't have a strong opinion, but probably two breaks would improve this code a bit

**Review by crsib — APPROVED** [review]
There is one place that looks to be typo, otherwise no major blockers for me

**Comment by dozzzzer** [comment]
@Paul-Licameli There's an issue with the **Change Pitch** effect:

- with a **mono track** it crashes Audacity when applying OR previewing the effect (with high-quality stretching both ON and OFF):
[Audacity_dbgrpt-6613-20230726T123238.zip](https://github.com/audacity/audacity/files/12171307/Audacity_dbgrpt-6613-20230726T123238.zip)

- with a **stereo track** it crashes Audacity only when previewing the effect and only with high-quality stretching ON: 
[Audacity_dbgrpt-6940-20230726T124710.zip](https://github.com/audacity/audacity/files/12171400/Audacity_dbgrpt-6940-20230726T124710.zip)

File: src/menus/SelectMenus.cpp (hunk 1 above)
Comment (Paul-Licameli) [reply src/menus/SelectMenus.cpp:41]: Ouch yes

File: libraries/lib-effects/MixAndRender.cpp (hunk 2 above)
Comment (Paul-Licameli) [reply libraries/lib-effects/MixAndRender.cpp:22]: Curious, why this one in particular?

**Comment by Paul-Licameli** [comment]
I pushed the small fix for Zero crossings:  it should examine selected tracks only.

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli There's an issue with the **Change Pitch** effect:
> 
>     * with a **mono track** it crashes Audacity when applying OR previewing the effect (with high-quality stretching both ON and OFF):
>       [Audacity_dbgrpt-6613-20230726T123238.zip](https://github.com/audacity/audacity/files/12171307/Audacity_dbgrpt-6613-20230726T123238.zip)
> 
>     * with a **stereo track** it crashes Audacity only when previewing the effect and only with high-quality stretching ON:
>       [Audacity_dbgrpt-6940-20230726T124710.zip](https://github.com/audacity/audacity/files/12171400/Audacity_dbgrpt-6940-20230726T124710.zip)
> 
> 
> _UPD: same behavior was observed when using Change Tempo effect_

Thanks.

Just tell me the symptoms -- I think these .zip files are not useful.

**Comment by dozzzzer** [comment]
@Paul-Licameli also a crash when applying **Sliding Stretch** on a mono track

**Comment by Paul-Licameli** [comment]
Never mind bisection -- I figured out all the reported problems and pushed two more commits.

**Comment by Paul-Licameli** [comment]
@crsib review the last 3 commits please?  Then I will rebase with some squashing and await complete testing.

**Review by crsib — COMMENTED** [review]
Last 3 commits reviewed!

**Comment by Paul-Licameli** [comment]
> Last 3 commits reviewed!

Metci

**Comment by Paul-Licameli** [comment]
Rebased onto recent master with squashing of fixups, otherwise no changes.

**Comment by Paul-Licameli** [comment]
> @Paul-Licameli some new issues were found:
> 
>     * destructive application of the **Compressor** effect both on mono and stereo tracks causes Audacity to crash
> 
>     * cutting/copying a region from a mono track and pasting it into a stereo track causes Audacity to crash
> 
>     * recording into an existing mono/stereo track (also Punch and Roll recording) causes Audacity to crash
> 
>     * **Plot Spectrum** always returns a blank frequency spectrum grid

New commits to address all of these problems.  Please review again, @crsib

**Comment by Paul-Licameli** [comment]
Rebased onto recent master (just one new commit there)

**Review by crsib — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
One more small commit.  Please review again @crsib

**Review by crsib — APPROVED** [review]

**Comment by dozzzzer** [comment]
Re-tested:

- destructive application of the Compressor effect both on mono and stereo tracks 
- cutting/copying a region from a mono track and pasting it into a stereo track
- recording into an existing mono/stereo track (also Punch and Roll recording)
- Plot Spectrum

**Pull request merged**